gh-103295: expose API for writing perf map files

gsallam opened the pull request:
#96123 added support for CPython to write /tmp/perf-<pid>.map files, associating instruction address ranges with a human-readable frame name for the Linux perf profiler.

Two external Python JIT compilers, [Cinder](https://github.com/facebookincubator/cinder/blob/cinder/3.10/Jit/perf_jitdump.cpp#L177-L185) and [Pyston](https://github.com/pyston/pyston/blob/dee48edd457518cf3d2cb93648f9cffe8aad6bce/Python/aot_ceval_jit.c#L5782-L5796), both also independently write to perf map files.

Since perf map files are one-per-process, multiple separate libraries trying to write perf map entries independently can lead to file corruption from simultaneous writes.

It's unlikely for both Cinder and Pyston JITs to be used in the same process, but it's quite reasonable to use one of these JITs along with CPython's native perf trampoline support.

This PR add a C-API to write to the perf map file. It also update the perf trampoline to use the new API to write to the perf map file. 

<!-- gh-issue-number: gh-103295 -->
* Issue: gh-103295
<!-- /gh-issue-number -->

gsallam committed:
Update osmodule.h

gsallam committed:
Update posixmodule.c

gsallam committed:
Update pycore_ceval_state.h

gsallam committed:
Update perf_trampoline.c

gsallam committed:
Update pylifecycle.c

gsallam committed:
Update pycore_ceval.h

gsallam committed:
Update _testinternalcapi.c

gsallam committed:
Create test_perfmaps.py

gsallam committed:
Update posixmodule.c

gsallam committed:
Update osmodule.h

gsallam committed:
Update osmodule.h

gsallam committed:
Update osmodule.h

gsallam committed:
Update posixmodule.c

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

gsallam committed:
Update osmodule.h

gsallam committed:
Update test_perfmaps.py

gsallam committed:
Update ignored.tsv

czardoz committed:
Fix trampoline APIs on windows, return int instead of void* from the init function

czardoz committed:
fix restructuredtext literals

czardoz committed:
add perfmaps docs under utilities

czardoz committed:
make "state" a non-required part of trampoline

czardoz committed:
make the API no-op on windows, instead of raising errors

pablogsal commented on Doc/c-api/perfmaps.rst:
```diff
@@ -0,0 +1,44 @@
+   Extensions are encouraged to directly call this API when needed, instead of
+   separately initializing the state by calling :c:func:`_PyOS_PerfMapState_Init`.
+
+.. c:function:: int _PyOS_PerfMapState_Fini(void)
```
Let's use the PEP 689 nomenclature for these APIs

pablogsal replied:
I also would prefer to have ALL of them under the `PyUnstable_` prefix for the time being until we understand more about the usage of the feature.

pablogsal commented on Modules/posixmodule.c:
```diff
@@ -16733,6 +16733,71 @@ INITFUNC(void)
     return PyModuleDef_Init(&posixmodule);
 }
 
+static PerfMapState perf_map_state;
```
I don't think this is a good place for this API, given that the other public `trampoline` APIs are in the sys module:

https://docs.python.org/3.12/library/sys.html#sys.activate_stack_trampoline

Let's move this one to the sys module as well.

gpshead commented on Python/perf_trampoline.c:
```diff
@@ -193,75 +193,28 @@ typedef struct trampoline_api_st trampoline_api_t;
-    fprintf(method_file, "%p %x py::%s:%s\n", code_addr, code_size, entry,
-            filename);
-    fflush(method_file);
+    char perf_map_entry[500];
```
I tend to not like large things allocated on the stack. When code runs with small stack sizes (as is common in threads started by non-Python C/C++ code in environments I've worked in), this kind of thing can bump into the limit at inopportune times.

I understand why `*_WritePerfMapEntry` wants to take entry_name as a single pre-constructed string for other outside of the CPython runtime uses cases (otherwise I'd suggest passing the utf-8 qualname and filename down directly for formatting in place).

I suggest: malloc and free a temporary buffer around the call to avoid a large buffer on the stack.

gsallam committed:
use PyUnstable as a prefix

gsallam committed:
move the implementation of the perf-map api to the sys module

gsallam committed:
Remove the perf-map API from the posixmodule.c

gsallam committed:
Update perfmaps.rst to reflect the PyUnstable prefix

gsallam committed:
Update perf_trampoline.c to use PyUnstable_WritePerfMapEntry

gsallam committed:
Update pylifecycle.c to use PyUnstable_PerfMapState_Fini

gsallam committed:
expose PyUnstable_PerfMapState_Fini

gsallam committed:
use assertIn instead of assertEqual and tear down the perf_map_state at the end of the test

gsallam committed:
use malloc instead of fixed buffer size

gsallam committed:
Fix perf map entry to have the START and SIZE as hex numbers without 0x

gsallam committed:
Add (uintptr_t) casting for the code_addr

gsallam committed:
update the perf map entry example to be hex without 0x

gsallam committed:
Update the doc with the PyUnstable prefix

czardoz committed:
Merge remote-tracking branch 'upstream/main'

czardoz committed:
change function name to the new one

gsallam committed:
fix tests_perfmaps

gsallam committed:
Update ignored.tsv

gsallam committed:
remove extra new line

gsallam committed:
remove extra trailing white space

gpshead committed:
check perf_map_entry for NULL.

nothing we can do but skip writing an entry, but at least _we_ won't be the cause of a crash and something that might generate an appropriate error real soon can be instead.

gpshead commented on Doc/c-api/perfmaps.rst:
It seems worthwhile specifically mentioning that the GIL must be held while calling these APIs.

Otherwise there might be some confusion as they mention that they are thread safe with their own lock, which is true, but I don't want anyone to assume that means the GIL does _not_ apply.

carljm replied:
Part of the goal here (and the reason for the separate lock) was to not require the GIL, as e.g. a JIT may want to do multithreaded compilation without holding the GIL, if it can compile without modifying Python objects. (Cinder JIT does this.)

Is there a risk you have in mind that would require the GIL to be held when using this API?

gpshead replied:
Upon further code inspection, it appears that only PyUnstable_PerfMapState_Init() requires the GIL (due to the PyErr_Set calls).  Just document that.

Agreed that you don't _want_ a GIL requirement given the purpose, I think the important one is WritePerfMapEntry and that one appears to be, as intended, fine for use w/o the GIL.

carljm replied:
Hmm, thanks for highlighting this point! The wrinkle is that `PyUnstable_WritePerfMapEntry` can call `PyUnstable_PerfMapState_Init` (if the perf map state is not already initialized.) We'd rather not force all callers to explicitly call `PerfMapState_Init` themselves.

We could just document that if you are going to call `WritePerfMapEntry` without the GIL (likely a less common case?), then you have to first ensure that `PerfMapState_Init` has already successfully been called in the process (with the GIL).

But I think our preferred option here is to stop setting Python exceptions in `PyUnstable_PerfMapState_Init` entirely, and instead make the entire API lower level and GIL-free. Instead we'd just return one of two error codes from `PerfMapState_Init`: one for "failed to open perf map file" and the other for "failed to allocate lock." If the "failed to open" code is returned, the caller can check `errno` themselves if they need more details.

Any thoughts/concerns about this approach? If it sounds OK, we can update the PR today.

gpshead replied:
Returning error codes instead of setting exceptions makes sense to me.

It is still worth documenting that these APIs are intended to have no GIL requirement for others future reference.

gpshead commented on Misc/NEWS.d/next/C API/2023-04-14-23-05-52.gh-issue-103295.GRHY1Z.rst:
perhaps add a ReST internal link to the howto/perf_profiling.rst doc?

gpshead commented on Doc/c-api/perfmaps.rst:
I suggest linking to this with a one or two sentence mention from the main [howto/perf_profiling.rst doc.](https://github.com/python/cpython/blob/main/Doc/howto/perf_profiling.rst) doc as that'll be the one people are most likely to read first.

czardoz committed:
Merge remote-tracking branch 'upstream/main'

czardoz committed:
update docs

czardoz committed:
remove rst extension from inline doc links

sunmy2019 commented on Include/internal/pycore_ceval.h:
```diff
@@ -66,7 +66,7 @@ typedef struct {
     void (*write_state)(void* state, const void *code_addr,
                         unsigned int code_size, PyCodeObject* code);
     // Callback to free the trampoline state
-    int (*free_state)(void* state);
```
Why is this change necessary? I see the only usage of this is set to `NULL` as a placeholder, and it is never called. (No code matches the pattern `free_state\(.*` now)


Is it used to support the user-customized function? Buy why removing the `(void* state)` if you need a `void *init_state(void)`?

sunmy2019 commented on Include/internal/pycore_ceval_state.h:
```diff
@@ -27,7 +27,7 @@ struct trampoline_api_st {
     void (*write_state)(void* state, const void *code_addr,
                         unsigned int code_size, PyCodeObject* code);
-    int (*free_state)(void* state);
+    void (*free_state)(void);
```
ditto

czardoz committed:
fix the free_state function signature

czardoz committed:
Merge remote-tracking branch 'upstream/main'

czardoz committed:
trim extra whitespace

gpshead reviewed (approved):
overall LGTM, thanks everybody.  updating the docs to mention the GIL requirement on the _Init method would be good.  and explicitly mentioning that the GIL is _not_ required for WritePerfMapEntry calls seems good as that is one of the main points of the API.

but those are docs changes and can be tweaked later as necesary.

sunmy2019 commented on Python/perf_trampoline.c:
```diff
@@ -491,10 +449,6 @@ _PyPerfTrampoline_Fini(void)
         tstate->interp->eval_frame = NULL;
     }
     free_code_arenas();
-    if (trampoline_api.state != NULL) {
-        trampoline_api.free_state(trampoline_api.state);
-        trampoline_api.state = NULL;
-    }
```
How about using the same thing as the init?
```c
    if (trampoline_api.state != NULL && trampoline_api.free_state != NULL) {
        trampoline_api.free_state(trampoline_api.state);
        trampoline_api.state = NULL;
    }
```

czardoz replied:
Keeping in mind that this particular backend for the trampoline API explicitly sets free_state to NULL, I don't quite see a benefit to spending cycles checking it. Can you elaborate on why this is better?

In any case, this will be a no-op, so I don't mind doing this change if there's a benefit to it :)

sunmy2019 replied:
If the `init_state` is optionally called, why should we not also call the `free_state`? I don't see why we should have this asymmetry.

If both `init_state` and `free_state` are no-op, maybe we can remove both from `trampoline_api`? 

----
`free_state` is now an unused variable. I think either we should use it, or we should remove it.

czardoz replied:
That's an interesting suggestion. Just to make sure I understand correctly - this involves changing the `trampoline_api_st` struct, which is a member of the `_ceval_runtime_state` struct?

Going by the [discussion in the initial implementation](https://github.com/python/cpython/pull/96123#pullrequestreview-1079722071), this API was designed to support different profiling backends, with the `perf` tool being one of them. In the future, I expect we might use this struct to support profiling tools that do require some state initialization and cleanup, for which this is useful to have. 

Changing the `trampoline_api_st` probably warrants a separate discussion. The scope of this particular PR is limited to:
1. Introducing C-APIs for managing the perf-map file.
2. Changing the `perf_trampoline.c` implementation to use the above APIs

Maybe @pablogsal or @gpshead might have more comments on this :)

czardoz replied:
Forgot to mention: I share your concern about the asymmetry here, thanks for noticing. We could get rid of the `init_state` call here, since that function is always NULL for the `perf` backend (y)

gsallam committed:
Return error codes instead of setting exceptions in PyUnstable_PerfMapState_Init

gsallam committed:
Add a note that holding the GIL is not required for these APIs

gsallam committed:
Remove the init_state call since it is always NULL for the perf backend

gsallam committed:
trim extra whitespaces

carljm committed:
simplify link to perf map docs from perf profiling

carljm committed:
document return values, other doc tweaks

carljm committed:
move headers to sysmodule.h

carljm committed:
pass through init return value from write-entry

sunmy2019 commented on Python/perf_trampoline.c:
```diff
@@ -193,75 +193,33 @@ typedef struct trampoline_api_st trampoline_api_t;
+        filename = PyUnicode_AsUTF8(co->co_filename);
+    }
+    size_t perf_map_entry_size = snprintf(NULL, 0, "py::%s:%s", entry, filename) + 1;
+    char* perf_map_entry = (char*) malloc(perf_map_entry_size);
```
Shall we use `PyMem_RawMalloc` or `PyMem_RawFree` here?

sunmy2019 commented on Python/perf_trampoline.c:
```diff
@@ -193,75 +193,33 @@ typedef struct trampoline_api_st trampoline_api_t;
-    fflush(method_file);
+    snprintf(perf_map_entry, perf_map_entry_size, "py::%s:%s", entry, filename);
+    PyUnstable_WritePerfMapEntry(code_addr, code_size, perf_map_entry);
+    free(perf_map_entry);
```
ditto

carljm committed:
apparently :c:data:`errno` doesn't exist

carljm committed:
use PyMem_Raw* instead of malloc/free

carljm committed:
Merge branch 'main' into gsallam/main

* main: (30 commits)
  gh-103987: fix several crashes in mmap module (#103990)
  docs: fix wrong indentation causing rendering error in dis page (#104661)
  gh-94906: Support multiple steps in math.nextafter (#103881)
  gh-104472: Skip `test_subprocess.ProcessTestCase.test_empty_env` if ASAN is enabled (#104667)
  gh-103839: Allow building Tkinter against Tcl 8.7 without external libtommath (GH-103842)
  gh-85984: New additions and improvements to the tty library. (#101832)
  gh-104659: Consolidate python examples in enum documentation (#104665)
  gh-92248: Deprecate `type`, `choices`, `metavar` parameters of `argparse.BooleanOptionalAction` (#103678)
  gh-104645: fix error handling in marshal tests (#104646)
  gh-104600: Make type.__type_params__ writable (#104634)
  gh-104602: Add additional test for listcomp with lambda (#104639)
  gh-104640: Disallow walrus in comprehension within type scopes (#104641)
  gh-103921: Rename "type" header in argparse docs (#104654)
  Improve readability of `typing._ProtocolMeta.__instancecheck__` (#104649)
  gh-96522: Fix deadlock in pty.spawn (#96639)
  GH-102818: Do not call `PyTraceBack_Here` in sys.settrace trampoline.  (GH-104579)
  GH-103545: Add macOS specific constants for ``os.setpriority`` to ``os`` (#104606)
  gh-104623: Update macOS installer to SQLite 3.42.0 (GH-104624)
  gh-104619: never leak comprehension locals to outer locals() (#104637)
  gh-104602: ensure all cellvars are known up front (#104603)
  ...

carljm committed:
fix PyMem_Free to PyMem_RawFree
